presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java

```diff
@@ -603,4 +627,117 @@ public QueryContext getQueryContext()
+    }
+
+    private List<OperatorMemoryReservationSummary> getOperatorMemoryReservations()
+    {
+        ListMultimap<List<Integer>, OperatorContext> operatorContexts = ArrayListMultimap.create();
+        accept(new VoidTraversingQueryContextVisitor<Void>()
+        {
+            @Override
+            public Void visitOperatorContext(OperatorContext operatorContext, Void nothing)
+            {
+                operatorContexts.put(
+                        ImmutableList.of(operatorContext.getDriverContext().getPipelineContext().getPipelineId(), operatorContext.getOperatorId()),
+                        operatorContext);
+                return null;
+            }
+        }, null);
+        ImmutableList.Builder<OperatorMemoryReservationSummary> result = ImmutableList.builder();
+        for (Collection<OperatorContext> operators : operatorContexts.asMap().values()) {
+            OperatorContext lastContext = getLast(operators);
+            long totalOperatorMemoryReservationInBytes = 0;
+            List<DataSize> reservations = new ArrayList<>();
+            for (OperatorContext context : operators) {
+                long reservedTotalMemoryInBytes = context.getCurrentTotalMemoryReservationInBytes();
+                totalOperatorMemoryReservationInBytes += reservedTotalMemoryInBytes;
+                reservations.add(succinctBytes(reservedTotalMemoryInBytes));
+            }
+            reservations.sort(reverseOrder());
+            result.add(new OperatorMemoryReservationSummary(
+                    lastContext.getOperatorType(),
+                    lastContext.getPlanNodeId(),
+                    ImmutableList.copyOf(reservations),
+                    succinctBytes(totalOperatorMemoryReservationInBytes),
+                    getAdditionalOperatorInfo(lastContext)));
+        }
+
+        return result.build().stream()
+                .sorted(comparing(OperatorMemoryReservationSummary::getTotal).reversed())
+                .collect(toImmutableList());
+    }
+
+    private Optional<String> getAdditionalOperatorInfo(OperatorContext context)
+    {
+        if (!taskPlan.isPresent()) {
+            return Optional.empty();
+        }
+
+        if (context.getOperatorType().equals(HashBuilderOperator.class.getSimpleName())) {
+            Optional<JoinNode> planNode = findPlanNode(context.getPlanNodeId(), JoinNode.class);
+            if (!planNode.isPresent()) {
+                return Optional.empty();
+            }
+            String info = planNode.get().getType().toString() + ";";
+            if (planNode.get().getDistributionType().isPresent()) {
+                info += planNode.get().getDistributionType().get() + ";";
+            }
+            return Optional.of(info);
+        }
+
+        if (context.getOperatorType().equals(HashAggregationOperator.class.getSimpleName()) ||
+                context.getOperatorType().equals(AggregationOperator.class.getSimpleName())) {
```

Comment:
Should we also store the STEP information whether this was doing partial/intermediate/final aggregate?